### ISSUE TITLE
CMake finds build requirements' libraries before requirements'

### ISSUE THREAD

**Issue by theodelrieu** [conan-io/conan#4117]
Hello,

I have a package that requires `Boost/1.66.0`. This package is a test library, I use it in my main package as a build requirement.

So `Boost/1.66.0` becomes a build requirement of my main package (even if the test library package is already built!).

The issue is I have a requirement on `Boost/1.68.0` in my main package.

When configuring CMake, the `Boost/1.66.0` is found. I guess there is an ordering problem when setting `CMAKE_PREFIX_PATH` and `CMAKE_MODULE_PATH`.

Here is a minimal use-case: 
[conan-bug.zip](https://github.com/conan-io/conan/files/2672573/conan-bug.zip)

Using Conan v1.9

To help us debug your issue please explain:

- [x] I've read the [CONTRIBUTING guide](https://raw.githubusercontent.com/conan-io/conan/develop/.github/CONTRIBUTING.md).
- [x] I've specified the Conan version, operating system version and any tool that can be relevant.
- [x] I've explained the steps to reproduce the error or the motivation/use case of the question/suggestion.

**Comment by lasote** [comment]
Hi, we introduced some improvements in the builds requires graph management. Could you please verify what is the behavior in the 1.13?
In case the error is still there I would ask @memsharded to take a look. Thanks Theo!

**Comment by theodelrieu** [comment]
Hi, I tried again with Conan 1.13 and the attachment posted earlier and it still uses the wrong zlib version (it should require 1.12.1 but instead uses 1.2.9)

**Comment by memsharded** [comment]
Hi @theodelrieu 

Could you please check if this failure remain with the latest fixes of conan 1.13.2?
I'd say the priority shouldn't matter as a conflict should be raised if there are incompatible Boost libraries that get to your package, no matter if build-requires or not.

Thanks!

**Comment by theodelrieu** [comment]
Nope, still finding 1.2.9 with 1.13.2. I agree that this should be a conflict usually, but I'm not sure about that if `override=True` is given.

**Comment by memsharded** [comment]
Hi @theodelrieu 

I have checked this. It is working fine, at least "as designed".
Build-requires are "private" requirements, you cannot override them from downstream packages, only from profiles. The reason is that overriding is actually the default behavior, and every build_requires defined in a recipe is defined there because it intends to be isolated from the rest of the world, otherwise you define it in profiles.

The zlib/1.2.11 override, is trying to override zlib in the dependency tree of your "bug/0.1" package, so your package and its dependants links with zlib/1.2.11. But the package "bug/0.1" is not, it is not linking with zlib (it does not require zlib), it is only temporarily there because of the build require.

In other words, if zlib is going to be linked in the "bug" package, it can NOT be a build require, it should be a regular require.

**Comment by theodelrieu** [comment]
>But the package "bug/0.1" is not, it is not linking with zlib (it does not require zlib).

I don't understand, I have the following in the `bug` recipe:

```python
    def requirements(self):
        self.requires("zlib/1.2.11@conan/stable", override=True)
```

Have I missed a subtle behavior of `self.requires`?

**Comment by memsharded** [comment]
Yes, there is a big difference when using ``override``:

```python
self.requires("zlib/1.2.11@conan/stable", override=True)
```

This means that your package do NOT require zlib. zlib is not in the dependency graph. Only if zlib was in the dependency graph, it would be overriden to 1.2.11. 

The big question is why your ``build_requirement`` is propagating the usage of zlib to its consumer, it doesn't seem a good idea, looks dangerous.

**Comment by theodelrieu** [comment]
But how can I both require and override? Should I put two lines?

In a real use-case, I have a `test-helpers` package which needs Boost to build and run, I then use this `test-helpers` as a build requirement for every other package I'm developing. I think it is safe in that case, but there might be a better way of doing this.

**Comment by memsharded** [comment]
A normal require will automatically override upstream. So if your downstream package requires "Boost/1.69", that will be propagated and override upstream, or conflict for those that cannot be overriden, like build requires. If your ``test-helpers`` package requires in a public way (not private, not build-requires, ``Boost/1.65``,  obviously you cannot link your tests in your  package with 2 different versions of Boost (one coming from the application, and one coming from the build-requires). In this case, I think the best approach is to treat it as a primary conflict source, and then update ``test-helpers`` to the same ``Boost`` version as the main dependency graph, so they do not conflict.

**Comment by theodelrieu** [comment]
I was completely unaware of that! It makes sense, so I guess this issue can be closed.
